src/Nexus/Extensibility/DataSource/DataSourceController.cs

```diff
@@ -149,166 +174,110 @@ internal class DataSourceController(
+        var availabilities = new double[stepCount, _dataSources.Length];
 
-                newResources.Add(newResource);
-            }
+        for (int dataSourceIndex = 0; dataSourceIndex < _dataSources.Length; dataSourceIndex++)
+        {
+            var tasks = new List<Task>(capacity: stepCount);
+            var currentBegin = begin;
+            var dataSource = _dataSources[dataSourceIndex];
 
-            if (isModified)
+            for (int i = 0; i < stepCount; i++)
             {
-                catalog = catalog with
+                var currentEnd = currentBegin + step;
+                var currentBegin_captured = currentBegin;
+                var i_captured = i;
+
+                tasks.Add(Task.Run(async () =>
                 {
-                    Resources = newResources
-                };
+                    var availability = await dataSource.GetAvailabilityAsync(catalogId, currentBegin_captured, currentEnd, cancellationToken);
+                    availabilities[i_captured, dataSourceIndex] = availability;
+                }, cancellationToken));
+
+                currentBegin = currentEnd;
             }
-        }
 
-        // TODO: Is it the best solution to inject these additional properties here? Similar code exists in SourcesController.GetExtensionDescriptions()
-        // add additional catalog properties
-        const string DATA_SOURCE_KEY = "data-source";
-        var catalogProperties = catalog.Properties;
+            await Task.WhenAll(tasks);
 
-        if (catalogProperties is not null &&
-            catalogProperties.TryGetValue(DATA_SOURCE_KEY, out var _))
-        {
-            // do nothing
         }
 
-        else
+        // calculate average (but ignore double.NaN values)
+        var averagedAvailabilities = new double[stepCount];
+
+        for (int i = 0; i < averagedAvailabilities.Length; i++)
         {
-            var type = DataSource
-                .GetType();
+            var sum = double.NaN;
+            var count = 0;
 
-            var nexusVersion = typeof(Program).Assembly
-                .GetCustomAttribute<AssemblyInformationalVersionAttribute>()!
-                .InformationalVersion;
+            for (int j = 0; j < _dataSources.Length; j++)
+            {
+                var value = availabilities[i, j];
 
-            var dataSourceVersion = type.Assembly
-                .GetCustomAttribute<AssemblyInformationalVersionAttribute>()!
-                .InformationalVersion;
+                sum = (sum, value) switch
+                {
+                    /* NaNs everywhere */
+                    (double.NaN, double.NaN) => double.NaN,
 
-            var repositoryUrl = type
-                .GetCustomAttribute<ExtensionDescriptionAttribute>(inherit: false)!
-                .RepositoryUrl;
+                    /* Replace sum with value */
+                    (double.NaN, _) => value,
```

Comment:
`NaN` is a valid final value for `sum`. It means that the data source has no idea how to calculate an availability. This could be the case for data sources which combine data from resources of other data sources. These data sources do not know anything about the availability of the original resources and currently (before this PR), they simply return `1`. But now that we have a pipeline, a value of 1 would render the combined pipeline availability a bit useless. So I think these types of data sources should return `NaN` instead. When they are the only data source in the pipeline, the return value is `NaN`.

That is why I think we should not initialize sum with `0`.

But in general, the availability plot seems to be broken sometimes and maybe this is caused by NaNs. #113 tracks this problem.

Reply:
Okay, than initializing with 0 is not the best idea. :-)